crwatcher/status.go

```diff
@@ -42,48 +43,58 @@ const (
 )
 
 type CustomResourceStatus struct {
-	Phase              ResourcePhase   `json:"phase"`
-	Reason             ConditionReason `json:"reason,omitempty"`
-	Message            string          `json:"message,omitempty"`
-	LastUpdateTime     metav1.Time     `json:"lastUpdateTime,omitempty"`
-	LastTransitionTime metav1.Time     `json:"lastTransitionTime,omitempty"`
+	Release            *release.Release `json:"release"`
+	Phase              ResourcePhase    `json:"phase"`
+	Reason             ConditionReason  `json:"reason,omitempty"`
+	Message            string           `json:"message,omitempty"`
+	LastUpdateTime     metav1.Time      `json:"lastUpdateTime,omitempty"`
+	LastTransitionTime metav1.Time      `json:"lastTransitionTime,omitempty"`
 }
 
-// SetPhase takes a custom resource status and returns the updated status, without updating the resource in the cluster.
-func SetPhase(status CustomResourceStatus, phase ResourcePhase, reason ConditionReason, message string) map[string]interface{} {
-	status.LastUpdateTime = metav1.Now()
-	if status.Phase != phase {
-		status.Phase = phase
-		status.LastTransitionTime = metav1.Now()
-	}
-	status.Message = message
-	status.Reason = reason
-
+func (s *CustomResourceStatus) ToMap() (map[string]interface{}, error) {
 	var out map[string]interface{}
-	jsonObj, err := json.Marshal(&status)
+	jsonObj, err := json.Marshal(&s)
 	if err != nil {
-		return map[string]interface{}{}
+		return nil, err
 	}
 	json.Unmarshal(jsonObj, &out)
-	return out
+	return out, nil
+}
+
+// SetPhase takes a custom resource status and returns the updated status, without updating the resource in the cluster.
+func (s *CustomResourceStatus) SetPhase(phase ResourcePhase, reason ConditionReason, message string) *CustomResourceStatus {
```

Comment:
I liked this better as a pure function...

Reply:
It's still pure, the first argument is just syntax-sugared to the left of the function name ;)